### PR TITLE
Add biases to eval_forward and parallelize both forward passes

### PR DESCRIPTION
**Pull request by tripplyons**
I noticed that in the `eval_forward` method the biases for the output layer are never added to the output logits, so I fixed that.

I was also able to remove the slow for loops in both forward passes, instead replacing them with `torch.einsum`. I confirmed that they give the same output for the same weights and input, but feel free to verify yourself.

### CODE REVIEW

**Comment by tripplyons** [comment]
Here is a Colab notebook demonstrating that it gets a >10x speedup and gives the same outputs (after fixing the missing biases):

https://colab.research.google.com/drive/1kyX3M-VAjUk_UuByYVUUz4YZCVbCK1OT?usp=sharing

**Comment by ChuxiJ** [comment]
looks great!
Nice PR

**Comment by pbelcak** [comment]
Hi, thanks for suggesting. Currently there are some other candidate implementations in the `sparse` branch. Could you please base your PR off that branch and add your implementation into the if-else fork of `eval_forward` among `sparse`, `naive`, and `greedy`?

Then it's going to be consistent with other implementations and can be an easy merge.
